### PR TITLE
Migrate from deriving to ppx_deriving, allow building with 4.06.0 

### LINKED ISSUE #30: Replacing "deriving" with "ppx_deriving"
Links does not compile with the latest OCaml compiler (4.03), because the package "deriving" is not compatible with 4.03 (actually I think it is one of the dependencies of "deriving" that's causing the issue). 

The package "deriving" appears to be obsolete, and for future projects with Links it may very well be worthwhile replacing "deriving" with "ppx_deriving". My own project would benefit from this change as well as the OCaml Multicore backend will be released for 4.03+.

According to Sam switching from "deriving" to "ppx_deriving" is not an entirely mechanical process. Furthermore, the proposed change is likely to be thorough, and since it may benefit Links in general James suggested that we implement this change on the de facto master-branch "sessions". 

The question remains: _Which parts of Links are sensitive to this change?_

### LINKED ISSUE #299: OCaml 4.06 compatibility
Work is needed to make Links (and dependent libraries) compatible with OCaml 4.06  @dhil can give more detailed examples.

### PR DESCRIPTION
**Pull request by frank-emrich**
The follwoing contributions are in this PR:
C1) I migrated all uses of the old deriving preprocessor to the ppx_deriving and translated all customizations accordingly. I've replaced the Dump/Pickle serialisation of deriving by ppx_deriving_yojson (fixes #30)
C2) I wanted to see whether ocaml 4.06.0 improves on the problem with warnings explained below. Therefore, I've established compability with ocaml 4.06.0. To this end, I've patched the cgi library (which seems to be abandoned) and included the two files directly into our code. Since we want to get rid of cgi anyway (see #293, #313), this should be an acceptable interim soolution. I also fixed a minor problem that occurred when compiling with 4.06.0 (fixes #299)

Problems:
P1) If the build fails on your machine, this is probably caused by some bugs in jbuilder and ppx_deriving that were only fixed recently. I did "opam upgrade" after running into these problems, so I'm not sure which version exactly of these two fixed the issue. To be on the safe side, we could just require the current version of both packages in our opam file.

P2) Previously, deriving generated modules implementing the desired functionality, for example a module Show_foo if some type foo was annoted with deriving (Show). ppx_deriving generated functions instead, in the case of [@@deriving show] these would be show_foo : foo -> string and pp_foo : Format.formatter -> foo -> unit.
The former function behaves like expected and uses the latter function, which implements the actual formatting.

One problem arising from this is that while unused modules are okay, unused functions trigger warnings by the compiler. Thus, we are now seeing dozens of warnings about unused generated functions. This is not simply due to superfluos deriving annotations. For example, the pp_foo function above may be required by some other pp_ function, but the show_foo function may not be used at all, thus yielding a warning that cannot easily be fixed.

The reason why I switched to 4.06.0 is that it allows warnings to be disabled on a  more fine-grained level than previous versions. This is already used by ppx_deriving, which generates for example some [@ocaml.warning "-39"] attributes, which disable warnings about unneeded "rec". However, the fact that we still see so many warnings shows that they are not generated in all places where they should be. There seem to be some github  issues in the repo of ppx_deriving about this.

Since having a big amount of warnings is undesirable, I suggest we implement one of the following solutions:
S1) We could help fixing the actual problems in ppx_deriving. This seems to be non-trivial since ppx_deriving uses the meta-programming facilities of OCaml and the code isn't exactly understandable at first sight.
S2) We could add the necessary attibutes to disable the warnings in our own code. Either for individual types/functions, or by just disabling the two types of warnings that we currently see on a per-module basis.

I briefly tested whether the change from Dump/Pickle to Yojson caused any problems when serialising continuations  but you're more than welcome to check if your favorite Links feature still works as expected.

### CODE REVIEW

File: core/jbuild [hunk 1]
```diff
@@ -10,6 +10,6 @@
   (wrapped false)
   (modes (native))
   (flags (:standard -safe-string -dtypes -w Ae-44-45-60 -g -cclib -lunix -thread))
-  (libraries (str deriving unix safepass cgi base64 ANSITerminal linenoise
+  (libraries (str deriving yojson ppx_deriving_yojson.runtime unix safepass base64 ANSITerminal linenoise
               cohttp lwt websocket websocket-lwt.cohttp))
-  (preprocess (action (run deriving/deriving ${<})))))
+  (preprocess (pps (ppx_deriving.std ppx_deriving_yojson)))))
```
Comment (dhil) [inline core/jbuild]: Brilliant

File: links.opam [hunk 2]
```diff
@@ -56,7 +56,8 @@ remove: [
 depends: [
   "jbuilder" {build}
   "deriving" {build}
```
Comment (dhil) [inline links.opam]: I think `deriving` should be removed as a build dependency.

**Comment by frank-emrich** [comment]
@slindley I have verified that using yojson fixes #124 and the other examples work as well, in particular the formlet one. I've tested it with and without handlers being enabled

File: links.opam [hunk 3]
```diff
@@ -7,7 +7,7 @@ bug-reports: "https://github.com/links-lang/links/issues"
 license: "GPL-2"
 
 available: [
-  ocaml-version >= "4.04.0" & ocaml-version <= "4.06.0"
+  ocaml-version >= "4.04.0" & ocaml-version <= "4.06.1"
```
Comment (dhil) [inline links.opam]: We should be able to remove the upper bound now, i.e. I think it is save to have `ocaml-version >= "4.04.0`" now.

File: links.opam [hunk 4]
```diff
@@ -60,7 +60,7 @@ depends: [
   "base64"
   "linenoise"
   "ANSITerminal"
-  "lwt"
+  "lwt" {>= "3.1.0"}
```
Comment (dhil) [inline links.opam]: Why do you need to constrain LWT?? Do we have some other dependency that depends on 3.1.0?

File: core/cgi.ml [hunk 5]
```diff
@@ -0,0 +1,328 @@
+(*
```
Comment (dhil) [inline core/cgi.ml]: It seems `cgi.ml` has been readded. I thought we had gotten rid of it.

Comment (frank-emrich) [reply core/cgi.ml]: See P2) in my PR description:  tldr: The removal of cgi isn't merged in yet, so I patched cgi as an immediate solution to work with 4.06.0

File: links.opam (hunk 3 above)
Comment (frank-emrich) [reply links.opam]: done!

**Comment by frank-emrich** [comment]
I have now solved the problem with warnings. In some cases, some deriving show/eq/yojson annotations were indeed unnecessary. In those 5 modules where some warnings simply could not be fixed, I added module-wide annotations to suppress the following two types of warnings:
1) unused functions
2) unnecessary rec flags in let 

From my side, this is now ready for merging

**Comment by dhil** [comment]
Cheers. I will have a look at this on Tuesday, when I am back in Edinburgh.

File: core/utility.ml [hunk 6]
```diff
@@ -33,9 +32,11 @@ let ( <| ) f arg = f arg
 
 (** {0 Maps and sets} *)
 module type OrderedShow = sig
-  type t
+  type t (**[@@deriving show]**)
   val compare : t -> t -> int
-  module Show_t : Deriving_Show.Show with type a = t
+  val show : t -> string
+  val pp : Format.formatter -> t -> unit
```
Comment (dhil) [inline core/utility.ml]: Remove this.

File: core/utility.ml [hunk 7]
```diff
@@ -33,9 +32,11 @@ let ( <| ) f arg = f arg
 
 (** {0 Maps and sets} *)
 module type OrderedShow = sig
-  type t
+  type t (**[@@deriving show]**)
```
Comment (dhil) [inline core/utility.ml]: Uncomment this.

File: core/types.ml [hunk 8]
```diff
@@ -571,17 +570,17 @@ let is_sessionable_row = is_sessionable_row IntSet.empty
...
 (* useful for debugging: types tend to be too big to read *)
 (*
-module Show_datatype = Show_unprintable (struct type a = datatype end)
-module Show_field_spec = Show_unprintable (struct type a = field_spec end)
-module Show_field_spec_map = Show_unprintable (struct type a = field_spec_map end)
-module Show_row_var = Show_unprintable (struct type a = row_var end)
-module Show_row = Show_unprintable (struct type a = row end)
-module Show_meta_type_var = Show_unprintable (struct type a = meta_type_var end)
-module Show_meta_row_var = Show_unprintable (struct type a = meta_row_var end)
+let pp_datatype = fun _ _ -> ()
```
Comment (dhil) [inline core/types.ml]: Should print ".."

**Pull request merged**